sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,175 @@
+# Copyright Red Hat 2021, Trevor Benson <trevor.benson@gmail.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import json
+import os
+import pexpect
+import shutil
+import subprocess
+
+from sos.collector.transports import RemoteTransport
+from sos.collector.exceptions import (ConnectionException,
+                                      ConnectionTimeoutException)
+from sos.utilities import (is_executable,
+                           sos_get_command_output_json)
+
+
+class SaltStackMasterTransport(RemoteTransport):
+    """
+    A transport for collect that leverages SaltStack's Master Pub/Sub
+    functionality to send commands to minions.
+
+    This transport will by default assume the use cmd.shell module to
+    execute commands on the minions.
+    """
+
+    name = 'saltstack'
+
+    def run_salt(self, cmd, **kwargs):
+        """Format and run a command with `salt` on the remote host.
+        """
+        return sos_get_command_output_json(
+            "salt --out json --static %s %s" % (self.address, cmd),
+            **kwargs
+        )
```

Comment:
Agreed. Intention was to place this into the `salstack.py` transport. Added below original code only for initial tests. I left it there ATM so I could get the WIP PR opened without delay, instead of moving it to the transport and another round of local tests.

> That does in turn raise the question of why this transport uses --out json, just to convert that json output back to a string before anything else is done to it?

The saltstack transports (salt master) output does not match the **control_persist** transport (ssh).

Without `--out json`  (ie.`salt store-0 cmd.shell 'hostname'`) returns:
```
store-0:
    EL8_store-0
```

- The first line is identical to the `self.address + ':'`.
- The second line contains the hostname. 

This felt like a poor match comparing to **control_persist** transport. Including `--out json` produces the json string. After conversion it provides identical output to **control_persist**. 

With `--out json` (ie. `salt --out json --static store-0 cmd.shell 'hostname'`) returns:

```
{
    "store-0": "EL8_store-0"
}
```

And of course `salt store-0 cmd.shell 'cat /etc/os-release'` returns:

```
store-0:
    NAME="Rocky Linux"
    VERSION="8.5 (Green Obsidian)"
    ID="rocky"
    ID_LIKE="rhel centos fedora"
    VERSION_ID="8.5"
    PLATFORM_ID="platform:el8"
    PRETTY_NAME="Rocky Linux 8.5 (Green Obsidian)"
    ANSI_COLOR="0;32"
    CPE_NAME="cpe:/o:rocky:rocky:8:GA"
    HOME_URL="https://rockylinux.org/"
    BUG_REPORT_URL="https://bugs.rockylinux.org/"
    ROCKY_SUPPORT_PRODUCT="Rocky Linux"
    ROCKY_SUPPORT_PRODUCT_VERSION="8"
```

To eliminate differences in data returned as the root cause for transport not working I created the `sos_get_command_output_json()` function. I am open to doing this another way as this was the first time I've looked at sos collect code so I may have gone a complex or unnecessary route.

Reply:
On this note, I'd recommend doing something like this in the saltstack transport, and dropping `sos_get_command_output_json()` from utilities:

```python

def _convert_output(self, json_output):
    return list(json.loads(json_output).values())[0]


def run_command(self, cmd, timeout=180, need_root=False, env=None,
                    get_pty=False):
    ret = super(SaltStackMasterTransport, self).run_command(cmd, timeout, need_root, env, get_pty)
    try:
        ret['output'] = self._convert_output(ret['output'])
    except Exception:
        # not json somehow?
        pass
    return ret
```

You can then drop `run_salt()` and leverage `run_command()` for everything (aside from the check against different salt binaries for support of the transport). This gives the benefit of always ensuring that the timeouts are handled along with sudo/root needs, pty obtaining, etc... it also makes debugging easier to trace through a single flow (ultimately thru `pexpect.spawn()` instead of `subprocess.Popen()`.)

Having both `run_salt()` and `run_command()` is not in and of itself a problem - `OCTransport` does something similar. However, that is so that there is some controlled difference between commands being executed on remote nodes (in pods) vs locally executed commands for various transport functions.

Once `run_command()` has this in place, we should be able to drop the `_read_file()` definition and use the base method that `control_persist` also uses. If we still have problems at that point, we at least know the issue is in the presentation of the output, not obtaining it.

Reply:
Implemented this change. Removed all other changes except the required import statement and adding the transport in the **TRANSPORTS** dictionary of `sosnode.py` file.